### PR TITLE
Add Vector3D and RotationMatrix Classes and Tests

### PR DESCRIPTION
**Pull request by julianmateu**
-Add a Vector 3D class with methods to add, multiply by scalar, normalize, etc.
-Add a Rotation Matrix class with several constructors.
-Add methods to normalize and conjugate Quaternions.
-Add tests for all those methods.

The tests can be improved by adding more cases and testing some interesting properties. There are still some methods to add, like rotate a vector using a Rotation Matrix or a Quaternion, or compose rotations multiplying their matrices, etc.

@basicNew

### CODE REVIEW

File: src/ProjectTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * Assert if two Quaternions are equal within a tolerance.
+     *
+     * @param expected the expected Quaternion.
+     * @param actual   the actual Quaternion.
+     * @param delta    the tolerance.
+     */
+    public void assertEqualsQuaternion(Quaternion expected, Quaternion actual, double delta) {
```
Comment (basicNew) [inline src/ProjectTest.java:22]: No need to change it, but just FYI you can write your custom assertions: https://ccguyka.wordpress.com/2015/02/08/custom-junit-assertions/

File: src/Quaternion.java [hunk 2]
```diff
@@ -83,6 +135,84 @@ public Quaternion multiply(double scalar) {
...
+        result = result.add(k.multiply(this.multiply(k)));
+        return result.multiply(-0.5);
+    }
+
+    /**
+     * Method to calculate the norm of a Quaternion.
+     *
+     * @return Returns the norm of the quaternion.
+     */
+    public double norm() {
```
Comment (basicNew) [inline src/Quaternion.java:180]: +1 on norm() vs normalize()!

File: src/Quaternion.java [hunk 3]
```diff
@@ -83,6 +135,84 @@ public Quaternion multiply(double scalar) {
...
+     * Method to normalize a Quaternion.
+     *
+     * @return Returns a new Quaternion with unit norm, un the same direction.
+     */
+    public Quaternion normalize() {
+
+        double norm = this.norm();
+
+        if (norm < MIN_TOLERANCE) {
+            throw new IllegalArgumentException("Trying to normalize a zero norm quaternion");
```
Comment (basicNew) [inline src/Quaternion.java]: +1 on the exception, not sure though that `IllegalArgumentException` is the one to be throwing now as this method doesn't have any arguments.

File: src/RotationMatrix.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            for (int j = 0; j < SIZE; j++) {
+                result[i] += this.getElement(i, j) * vector.getElement(j);
+            }
+        }
+
+        return new Vector3D(result);
+    }
+
+    /**
+     * Method to assert if to Rotation Matrices are equal.
```
Comment (basicNew) [inline src/RotationMatrix.java]: to -> two

File: src/RotationMatrix.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @return Returns True if they are equal, and False otherwise.
```
Comment (basicNew) [inline src/RotationMatrix.java]: AFAIK you use uppercase to describe the classes and lowercase to describe the instances (hence the comment would be "Returns true if they are equal, and false otherwise"). But I'm not sure what is the standard for Java comments.

File: src/RotationMatrix.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    //************ Getters & Setters ************//
+
+    /**
+     * Method to get the element i,j of the matrix.
+     *
+     * @return Returns the element in the ith row and jth column of the matrix.
+     */
+    public double getElement(int i, int j) {
+        if (i < 0 || i >= SIZE || j < 0 || j >= SIZE) {
+            throw new IllegalArgumentException("Attempting to access an inexistent element");
```
Comment (basicNew) [inline src/RotationMatrix.java]: This time the exception is right, but maybe we could improve the error message. How about adding the `i` and `j` to the  message and stating the matrix size (e.g. "Element out of bounds (4,3). Matrix size is 3x3")?.

File: src/TestQuaternion.java [hunk 7]
```diff
@@ -41,7 +27,31 @@ public void testEquals() {
         assertEquals(a, new Quaternion(1, 0, -1, 5));
         assertEquals(a, a);
         assertFalse(a.equals(new Quaternion(-1, 0, 0, 5)));
+    }
 
+    /**
+     * Tester for the constructors.
+     */
+    public void testConstructors() {
+        double tol = 1e-10; // Tolerance.
```
Comment (basicNew) [inline src/TestQuaternion.java]: Abbreviating names is generally not a good practice (In particular if you have to write a comment to explain it :)).

File: src/TestQuaternion.java [hunk 8]
```diff
@@ -192,4 +202,95 @@ public void testMultiplicationQuaternion() {
...
+    public void testNorm() {
+
+        double tol = 1e-6; // Tolerance.
+        assertEquals(new Quaternion(1, 1, 1, 0).norm(), Math.sqrt(3), tol);
+        assertEquals(new Quaternion(1, 0, 0, 0).norm(), 1, tol);
+        assertEquals(new Quaternion(-1, 1, -1, 1).norm(), 2, tol);
+        assertEquals(new Quaternion(1, 2, -3, 4).norm(), Math.sqrt(30), tol);
+        assertEquals(new Quaternion(2, 2, 2, 2).norm(), 4, tol);
+
+        try {
```
Comment (basicNew) [inline src/TestQuaternion.java:245]: +1 on testing the exception! Note however that there are some built-in ways of doing that in JUnit http://stackoverflow.com/questions/15216438/junit-testing-exceptions

File: src/TestRotationMatrix.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    /**
+     * Constructor
+     */
+    public TestRotationMatrix(String name) {
+        super(name);
+    }
+
+    /**
+     * Tester for the equals() method.
+     */
```
Comment (basicNew) [inline src/TestRotationMatrix.java]: IMO there is no need for this sort of comment if your test is called `testEquals`.

File: src/TestRotationMatrix.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+        assertEquals(a, new RotationMatrix(90, axis));
+        assertEquals(a, a);
+        assertFalse(a.equals(new RotationMatrix(45, axis)));
+
+    }
+
+    /**
+     * Tester for the getElement method.
+     */
```
Comment (basicNew) [inline src/TestRotationMatrix.java]: Same as above

File: src/TestRotations.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
+//: object/TestRotations.java
+
+import junit.framework.*;
+
+/**
+ * JUnit Class to test the rotation methods from the Quaternion and RotationMatrix classes.
+ *
```
Comment (basicNew) [inline src/TestRotations.java:7]: Maybe instead of having this what you should do is testing the rotation in quaternion with a matrix and the other way around :)

**Comment by basicNew** [comment]
Looks great, just some minor comments

File: src/Quaternion.java (hunk 3 above)
Comment (julianmateu) [inline src/Quaternion.java]: `ArithmeticException` perhaps ?

Comment (basicNew) [inline src/Quaternion.java]: Yes. Or you can create your own new exception class :)

File: src/TestRotations.java (hunk 11 above)
Comment (julianmateu) [inline src/TestRotations.java:7]: I don't get it, you mean rotating first with a quaternion and then "unrotate" with a matrix and see if I return to the original vector?

Comment (basicNew) [inline src/TestRotations.java:7]: Or rotating the same vector with a quaternion and a matrix and check that the result is the same. My point is that when you write a unit test you basically want to see if the behavior you implemented is right. If you happen to have another object that does the same, but with a different implementation, it is a great candidate helper to write your test.

Comment (julianmateu) [inline src/TestRotations.java:7]: And that is what I did, I think:

```
        v = new Vector3D(1, 0, 0);
        axis = new Vector3D(1, 1, 1);
        q = new Quaternion(30, axis);
        m = new RotationMatrix(30, axis);
        assertEqualsVector3D(q.rotate(v), m.rotate(v), tolerance);
```

Comment (basicNew) [inline src/TestRotations.java:7]: Sure, my point is that this shouldn't be in a separate `TestRotations` test, but it the `TestQuaternion` and `TestRotationMatrix` in the `testRotate` method

Comment (julianmateu) [inline src/TestRotations.java:7]: That wouldn't lead to duplication in the code? I would be testing that the rotation with a matrix gives the same result as the rotation with a quaternion in the `RotationMatrix` test, and that the quaternion rotation gives the same result as the matrix in the `Quaternion` test.

Comment (basicNew) [inline src/TestRotations.java:7]: That's a great question! The answer is that it depends; are the edge cases the same for both tests? So e.g. an edge case for the rotation matrix may be trying a matrix that has all zeroes, but that may not have sense for a quaternion. So  you have to always think what are the edge cases for the class that you are writing the tests.
